plugins/mac-capture/mac-audio.m

```diff
@@ -637,6 +645,214 @@ static bool coreaudio_init(struct coreaudio_data *ca)
 	return false;
 }
 
+static NSString *halPluginFileName =
+	@"/Library/Audio/Plug-Ins/HAL/OBSDesktopAudioCapture.driver";
+
+static bool is_hal_device_updated()
+{
+	NSDictionary *halPlist =
+		[[NSBundle bundleWithPath:halPluginFileName] infoDictionary];
+	NSString *halVersion =
+		[halPlist objectForKey:@"CFBundleShortVersionString"];
+	NSString *halBuild =
+		[halPlist objectForKey:(NSString *)kCFBundleVersionKey];
+
+	NSString *obsVersion = [[[NSBundle mainBundle] infoDictionary]
+		objectForKey:@"CFBundleShortVersionString"];
+	NSString *obsBuild = [[[NSBundle mainBundle] infoDictionary]
+		objectForKey:(NSString *)kCFBundleVersionKey];
+	return [halVersion isEqualToString:obsVersion] &&
+	       [halBuild isEqualToString:obsBuild];
+}
+
+static bool is_hal_device_installed()
+{
+	AudioDeviceID deviceId;
+	coreaudio_get_device_id(CFSTR(DESKTOP_CAPTURE_HAL_DEVICE_UID),
+				&deviceId);
+	return deviceId != 0;
+}
+
+static bool modify_hal_plugin(bool reinstall)
+{
+	NSString *removeCmd;
+	NSString *installCmd;
+	removeCmd = [NSString
+		stringWithFormat:@"rm -rf '%@' && ", halPluginFileName];
+	if (reinstall) {
+		NSString *halPluginSourcePath = [[NSBundle mainBundle]
+			pathForResource:@"OBSDesktopAudioCapture.driver"
+				 ofType:nil];
+		NSString *halPluginDestinationPath =
+			@"/Library/Audio/Plug-Ins/HAL/";
+		installCmd =
+			[NSString stringWithFormat:@"cp -R '%@' '%@' && ",
+						   halPluginSourcePath,
+						   halPluginDestinationPath];
+	} else {
+		installCmd = @"";
+	}
+	NSString *restartCmd =
+		@"launchctl kickstart -k system/com.apple.audio.coreaudiod";
+	NSString *appleScriptString = [NSString
+		stringWithFormat:
+			@"do shell script \"%@%@%@\" with administrator privileges",
+			removeCmd, installCmd, restartCmd];
+
+	NSAppleScript *appleScriptObject =
+		[[NSAppleScript alloc] initWithSource:appleScriptString];
+	NSDictionary *error;
+	NSAppleEventDescriptor *eventDescriptor =
+		[appleScriptObject executeAndReturnError:&error];
+	if (eventDescriptor == nil) {
+		const char *errorMessage = [[error
+			objectForKey:@"NSAppleScriptErrorMessage"] UTF8String];
+		blog(LOG_WARNING,
+		     "[mac-capture]: Failed to modify HAL plugin: %s",
+		     errorMessage);
+		return false;
+	}
+
+	// Hack to wait until CoreAudio has reloaded. I hate it too.
+	sleep(5);
```

Comment:
Is there any way to make this async? I can imagine that it might the UI code a bit messy because you'd have to add a signal to handle the eventual status of the audio subsystem after waiting for 5 seconds.

Reply:
Yeah, I think it'd be overkill.
While I didn't try, it's likely to be too much code for too little reward (and there's little purpose to making it async imo, since the next thing the user does after clicking the button is selecting the device, which only is available after the 5 seconds).